braid/braid_status.c

```diff
@@ -359,8 +359,13 @@ braid_StatusSetRFactor(braid_Status status,
       braid_Int          ilower   = _braid_GridElt(grids[level], ilower);
       braid_Int          ii       = index+1 - ilower;
 
-      /* Set refinement factor */
-      rfactors[ii] = rfactor;
+      /* Set refinement factor
+       * 
+       * Note: make sure index ii is positive.  There is one step with
+       * Richardson in FCRelax that is with a C-point from the left-neighboring
+       * processor.  Thus, ii can be negative. */
+      if( ii >= 0 )
```

Comment:
@jbschroder I think I will need an explanation for this one as well.  I didn't immediately see how or why `FCRelax` causes this problem.  I am also wondering how this (and the rest of the Richardson stuff) plays with the periodic feature.  Thanks again!

Reply:
Richardson needs to take a step with the C-point to the left of the left-most F-point (on processor).  And sometimes this C-point needs to come from the neighboring processor to the left.  In that case, ii can be negative, and the current processor doesn't "own" this point.  So, I figured that any rfactor requests on this point should be ignored.

Regarding periodic, that's a good point.  I didn't test that.  Is there a case with periodic where ii can be negative, but not the initial condition, and we might want to refine?

Reply:
Okay, I see where this happens in `FCRelax` (I think).  However, I don't think users should be setting rfactor for any of these cases, since it's based on the coarse step size.  Does this ever break the code, i.e., cause an out-of-bounds array issue or something?  In looking at this `braid_status.c` code, I'm not sure `ii` is ever negative since it's set as follows:

    braid_Int          ii       = index+1 - ilower;

Basically, we store the rfactors using the "right" index for the interval.

Reply:
We can chat on the phone, if that would be easier.  Just let me know. 

This extra Step() for Richardson happens in relax.c from lines 191 -- 214.  The corresponding communication (if needed, for this C-point corresponding to the left-most interval) is done above in lines 91 -- 120.

And, I agree that users should not be setting RFactor in this case.  One possible fix would be to set "level" to "level+1" in lines 201 -- 202 in FCRelax, so that the StepStatus structure is initialized for a coarse level.  This way, even if the user set rfactor for this step, it would be ignored, because (level == 1)  

Aside: I was getting an array out of bounds, because the time index being stepped from (line 201 relax.c) is (ci - cfactor - 1).  So even (index+1 - ilower) could be negative if cfactor was > 2.

Reply:
So, the proposed changes would be to set "level" to "level+1" in lines 201--202 in FCRelax, and to then also remove the check in SetRFactor "if( ii >= 0 )"

Reply:
This could be a little confusing to a user, if they are using a different spatial grid or time-stepper (or anything else) on the coarser grid, because Step does pass a "fine-level" vector to the user here.  But, this is a weird algorithmic step.  It's with a fine-level vector, but with a coarse-level dt.

Reply:
We will leave as is for now and create an issue to make it easier for users to know when it's okay to set rfactor.